rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,124 @@
+
+The [deploy-auto-update-changes.yml](https://github.com/gravitational/cloud/blob/master/.github/workflows/deploy-auto-update-changes.yml) workflow will need to be updated to include an additional job that updates the cloud-stable version of client tools.
+
+Unlike the agent version discovery model, there will not be an endpoint to identify if a critical update is available. The reason for this is because updates are not executed in a scheduled update window. Teleport client tools will check this endpoint everytime `login` or `status` is executed, and update accordingly if the cached version does not match the cloud-stable version.
+
+### Caching
+
+A single version of the Teleport client tools will be stored in the user's $HOME directory. Teleport already makes use of a .tsh directory for storing tsh config. Cached versions of tsh/tctl will also live within this directory under ` $HOME/.tsh/bin/{tsh,tctl}`. Whenever a new version of the client tools are available, the existing cached version will be replaced.
+
+The .tsh directory is only accessilbe to the user and we should make sure the new `bin` directory also only has user permissions.
+
+Everytime `login` or `status` command is executed, tsh will compare the currently cached version of the client tools with the current cloud-stable version of Teleport. If the cached version is behind or ahead of the current stable version, a new version of the client tools will be downloaded to replace the existing ones. When downloading the Teleport client tools, the binaries will be downloaded directly. We need to make sure to download a version that is compatible with the system.
+
+Because only a single version of the client tools will be cached at a time, we do not need to worry about cleaning up the cache.
+
+### Package upgrade
+
+As an alternative to caching, we've considered directly upgrading the Teleport packages.
+
+This approach will be difficult to implement because we support multiple package repositories and installation methods. The amount of maintenance and testing required to support this solution is probably not worth the effort at this time.
+
+Another downside to this approach would be that upgrades would require sudo/admin privileges.
+
+### Config
+
+Automatic updates for client tools can be configured through tsh configuration. If `DISABLE_AUTO_UPDATE=true`, then auto updates for client tools will be disabled. If `DISABLE_UPDATE_PROMPT=true`, then the user will not be prompted to confirm the update. Both values will default to false.
+
+### Binary Execution
+
+The client tools will need a way to execute the cached binaries. When tsh/tctl is executed, it will now need to decide whether to execute the command as usual, or fork a new process using the cached binary.
```

Comment:
I worry a little bit about this approach. If I install tsh 13.1.1 via `apt`, I might be very confused if I find that I'm running tsh 13.1.2 via `tsh version`. This could be especially frustrating when debugging. It could also lead to security issues, because the version of `tsh` shown in a package list may not be impacted by a CVE, but the version in `.tsh` that gets executed might be.

I wonder if we should offer `tsh-cloud` packages that provide a shim which always downloads the latest version, or alternatively, insist that Cloud users use strictly user-local/script-based installations of tsh/tctl.

On macos, `brew` might be considered prior-art as a CLI that follows an evergreen pattern for updates.

I'm not sold on any of these approaches.

@r0mant any thoughts?

Reply:
Yeah, none of these approaches seem like a perfect solution. Open to suggestions.

Reply:
When in doubt, look at how others do it :) Let's do a bit more research on prior art. For example:

- Terraform. Does it have automatic upgrades? I'm not sure it does, I think it only prints that update is available thus letting users upgrade themselves using whatever method they prefer?
- Tailscale has [tailscale update](https://tailscale.com/kb/1080/cli/#update-alpha) command. How does it work if you installed it via apt package? 

Either way, I agree that if we have tsh re-execute itself, we won't be able to get rid of this potential discrepancy between the package version and binary version. So we might have to have a shim script that can be versioned separately, yeah.

Reply:
I'm thinking something much simpler, a single version check that says "run tsh sync to sync to the right version" on failure. 

The number of people using teleport cloud with Linux workstations is probably pretty small compared to Mac. On Mac it should be safe for us just to overwrite the current binary and call it a day, no?

Reply:
### terraform update

- https://spacelift.io/blog/terraform-version-upgrade

Terraform uses the `tfenv` tool as a version manager. `tfenv` can be used to install multiple versions of terraform. `tfenv install 1.3.9` would download the tarball and install terraform in `$HOME/.tfenv/versions/1.3.9/terraform`. Then you can switch between versions with `tfenv use 1.3.9`. This sets the `TFENV_TERRAFORM_VERSION=1.3.9`.

### awsc cli

- https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html#getting-started-install-instructions

AWS CLI does not have any update command. Their docs simply instruct you to reinstall the tool when you're ready to update.

### tailscale

- https://tailscale.com/kb/1067/update/?q=update&tab=linux

The `tailscale update` command is supported on windows and some linux distros. The implementation looks like it checks the OS/distro and then uses the package manager to update the tool. On macOS they automatically update the tool using https://github.com/mas-cli/mas, but tailscale needs to be installed via the app store to support this feature.

Reply:
> I'm thinking something much simpler, a single version check that says "run tsh sync to sync to the right version" on failure.

We ended up with this idea of caching a binary on the user's $HOME directory because of two main reasons:
- it would not require root/admin privileges to update the tools
- it could take up a decent amount of work to test and maintain support for the various different installation methods

> On Mac it should be safe for us just to overwrite the current binary and call it a day, no?

If we want support for TouchID, the tsh.pkg needs to be installed which would then require root/admin privileges. It sounds like though this won't be a problem after this quarter.

Looking at those three examples, it looks like there is no perfect answer. I'll just go ahead and list these in the rfd and